### PR TITLE
feat: register signal handlers for graceful shutdown

### PR DESCRIPTION
**Pull request by olavloite**
Register signal handlers for SIGTERM, SIGINT, and SIGQUIT to allow graceful shutdown, and add support for the SMART shutdown mode from PostgreSQL. The SMART shutdown mode allows the server to drain connections and shut down after all connections have closed.

Registering signal handlers is a best-effort and requires that the `sun.misc.Signal` and `sun.misc.SignalHandler` classes are available on the JVM where PGAdapter is running.

Updates #2040
Updates #1189

### CODE REVIEW

File: src/main/java/com/google/cloud/spanner/pgadapter/Server.java [hunk 1]
```diff
@@ -40,15 +41,30 @@
...
 import java.util.Locale;
 import java.util.Objects;
 import java.util.concurrent.ThreadLocalRandom;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 import org.apache.maven.model.Model;
 import org.apache.maven.model.io.xpp3.MavenXpp3Reader;
+import org.codehaus.mojo.animal_sniffer.IgnoreJRERequirement;
+
+// import sun.misc.Signal;
```
Comment (hengfengli) [inline src/main/java/com/google/cloud/spanner/pgadapter/Server.java]: Remove?

File: src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+      shutdownHandler.shutdown(ShutdownMode.SMART);
+      // Wait until the server port is no longer open.
+      Stopwatch stopwatch = Stopwatch.createStarted();
+      while (isPortListening(proxyServer.getLocalPort())
+          && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 2000) {
+        Thread.yield();
+      }
+      assertEquals(State.STOPPING, proxyServer.state());
+      // Verify that we can still use the existing connection.
+      try (ResultSet resultSet = connection.createStatement().executeQuery(sql)) {
```
Comment (hengfengli) [inline src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java:93]: If this connection is not closed, pgadapter will not shut down under SMART mode, right?

File: src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+      }
+      // Verify that the server is still stopping, but not yet terminated.
+      assertEquals(State.STOPPING, proxyServer.state());
+    }
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    while (proxyServer.state() == State.STOPPING
+        && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 2000) {
+      Thread.yield();
+    }
+    // Verify that closing the last connection terminates the server.
```
Comment (hengfengli) [inline src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java:120]: So the last connection is closed at line 114, right?

File: src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      shutdownHandler.shutdown(ShutdownMode.FAST);
+      // Verify that the server stops.
+      Stopwatch stopwatch = Stopwatch.createStarted();
+      while (proxyServer.state() != State.TERMINATED
+          && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 2000) {
+        Thread.yield();
+      }
+      assertEquals(State.TERMINATED, proxyServer.state());
+
+      // Verify that existing connection has been invalidated.
```
Comment (hengfengli) [inline src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java:149]: So the existing connection gets the signal to terminate. Currently, FAST is almost same to IMMEDIATE, right?

**Review by hengfengli — APPROVED** [review]
LGTM. Left a few comments.

File: src/main/java/com/google/cloud/spanner/pgadapter/Server.java (hunk 1 above)
Comment (olavloite) [reply src/main/java/com/google/cloud/spanner/pgadapter/Server.java]: Oops. Removed.

File: src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java (hunk 2 above)
Comment (olavloite) [reply src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java:93]: That is correct. However, initiating a FAST or IMMEDIATE shutdown while a SMART shutdown is pending, will override the SMART shutdown and close PGAdapter.

File: src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java (hunk 3 above)
Comment (olavloite) [reply src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java:120]: Yes, correct. The end of the `try` block automatically closes the connection (also if the block is exited through for example an exception).

File: src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java (hunk 4 above)
Comment (olavloite) [reply src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java:149]: Correct. The only difference at this moment is that FAST will wait for up to 1 second for the Spanner client to close (so wait for the session pool to clean up all sessions that were in use). IMMEDIATE will just do a fire-and-forget close of the Spanner client.

**Pull request merged**